WzComparerR2.LuaConsole/Examples/DumpSounds.lua

```diff
@@ -0,0 +1,110 @@
+import 'WzComparerR2.PluginBase'
+import 'WzComparerR2.WzLib'
+import 'System.IO'
+import 'System.Xml'
+
+------------------------------------------------------------
+
+local function enumAllWzNodes(node) 
+  return coroutine.wrap(function()
+    coroutine.yield(node)
+    for _,v in each(node.Nodes) do
+      for child in enumAllWzNodes(v) do
+        coroutine.yield(child)
+      end
+    end
+  end)
+end
+
+local function isSound(value)
+  if value and type(value) == "userdata" and value:GetType().Name == 'Wz_Sound' then
+    return true
+  else 
+    return false
+  end
+end
+
+local p = Path.GetInvalidFileNameChars()
+local ivStr = ""
+for i, v in each(p) do
+  if v >= 32 then
+    ivStr = ivStr .. string.char(v)
+  end
+end
+local ivPattern = "["..ivStr.."]"
+------------------------------------------------------------
+
+-- all variables
+local topWzPath = 'Sound\\Bgm00.img'
+local topNode = PluginManager.FindWz(topWzPath)
+local outputDir = "D:\\wzDump"
+
+------------------------------------------------------------
+-- main function
+
+if not topNode then
+  env:WriteLine('"{0}" not loaded.', topWzPath)
+  return
+end
+
+
+
+-- enum all wz_images
+for n in enumAllWzNodes(topNode) do
+  local img = Wz_NodeExtension.GetNodeWzImage(n)
+  
+  if img then
+    --extract wz image
+    env:WriteLine('(extract)'..(img.Name))
+    if img:TryExtract() then
+    
+      local dir = outputDir.."\\"..(n.FullPathToFile)
+      local dirCreated = false
+      
+      --find all sound
+      for n2 in enumAllWzNodes(img.Node) do
+        local sound = n2.Value
+        if isSound(sound) then
+          
+          local fn = n2.FullPath:sub(img.Name:len()+2):gsub("\\", "."):gsub(ivPattern, "")
+          if not n2.Text:find("\\.") then
+            if sound.SoundType == Wz_SoundType.Mp3 then
+              fn = fn .. ".mp3"
+            end
+            if sound.SoundType == Wz_SoundType.WavRaw then
+              fn = fn .. ".wav"
+            end
+          end
+          fn = Path.Combine(dir, fn)
+          
+          --ensure dir exists
+          if not dirCreated then
+            if not Directory.Exists(dir) then
+              Directory.CreateDirectory(dir)
+            end
+            dirCreated = true
+          end
+          
+          --save sound
+          env:WriteLine('(output)'..fn)
+          local file = io.open(fn, "wb")
+          local bytes = sound:ExtractSound()
+          for i = 0, (sound.DataLength - 1) / 4096 do
+            file:write(string.char(table.unpack(bytes, i * 4096, math.min(i * 4096 + 4095, sound.DataLength - 1))))
+          end
+          file:close()
```

Comment:
```suggestion
          File.WriteAllBytes(fn, sound:ExtractSound())
```

https://github.com/Kagamia/WzComparerR2/issues/170#issuecomment-934323741

Really significant speed difference.

Reply:
in #170 we have unblocked passing byte[] as argument by contributing to NLua, suggest to use .net System.IO library instead of lua IO function.